### ISSUE TITLE
Corner case where routing fails.

### ISSUE THREAD

**Issue by sarvghotra** [quinnj/Endpoints.jl#4]
``` julia
using Endpoints, Base.Test, HTTP
Endpoints.@GET "test" ()->"testing GET resource"
Endpoints.@GET "test/{var::String}" var->"testing $var resource"
HTTP.serve(handler = Endpoints.handler)
```

In the above case if we browse http://127.0.0.1:8081/test/sarv then the routing is correct as expected. But for the below case it fails. Instead of displaying "testing sarv resource" it shows **"/test/sarv"** i.e when nothing matched.

```julia
using Endpoints, Base.Test, HTTP
Endpoints.@GET "test" ()->"testing GET resource"
Endpoints.@GET "test/{var::String}" var->"testing $var resource"
Endpoints.@GET "test/sarv/ghotra" var->"testing sarv ghotra conflict case" 
HTTP.serve(handler = Endpoints.handler)
```

The reason for the failure is that `PATH_LOOKUPS` has `sarv` as an entry due to this line - `Endpoints.@GET "test/sarv/ghotra" var->"testing sarv ghotra conflict case" `. So when we browse for "/test/sarv", the actual value of `vars_and_args` from [here](https://github.com/quinnj/Endpoints.jl/blob/master/src/Endpoints.jl#L152) is  `[Val{:test}, Val{:sarv}]` instead of `[Val{:test}, "sarv"]` (expected).

**Comment by quinnj** [comment]
Yeah, I started thinking  about this the other day. I think what we need to do is re-architect `PATH_LOOKUPS` so that it's not just a simple dict lookup, but that we can provide some kind of hierarchy. i.e. We would lookup `Val{:test}`, the move on to `"sarv"` and only lookup the next path segment based on the prefix of `Val{:test}`. I'm not sure the best data structure for this, but something to think about.

**Comment by sarvghotra** [comment]
I'll also try to come up with a work-around to this issue. 

Also, there are some other problems arising due to the multi dispatch approach. I'll compile a list of them and will share in a new issue, very soon.
